### PR TITLE
Macro `map_with_index`

### PR DESCRIPTION
**Pull request by Blacksmoke16**
* Implements the `map_with_index` method on `TupleLiteral` and `ArrayLiteral`.
* ~Reorganizes the macro method specs a bit~
  * ~Alphabetizes each type~
  * ~Use the constant type vs string~
  * ~Updates some type's structure to use the standard format~
    * ~Mainly just `TupleLiteral` and `ArrayLiteral`.~
  * ~Marks some missing tests as pending~
    * ~Can make another PR later to fill these in.~

Reorganization will be handled in another PR.

Use case is something like (pseudo code):
```cr
initializer = some_class_def.methods.find(&.name.==("initialize"))
service_ann.args.map_with_index { |arg, idx| arg.as(initializer.args[idx].restriction) } 
```

### CODE REVIEW

**Comment by straight-shoota** [comment]
Could you please separate reorganization and added feature into separate commits (or maybe even PRs)? Thanks!

**Comment by Blacksmoke16** [comment]
@straight-shoota I'll make another PR with the reorganization plus implementing the missing specs.

File: spec/compiler/macro/macro_methods_spec.cr [hunk 1]
```diff
@@ -588,6 +588,32 @@ module Crystal
         assert_macro "x", %({{x.map { |e| e.id }}}), [ArrayLiteral.new(["hello".call] of ASTNode)] of ASTNode, "[hello]"
       end
 
+      describe "#map_with_index" do
+        context "with both arguments" do
+          it "should return the resulting array" do
```
Comment (straight-shoota) [inline spec/compiler/macro/macro_methods_spec.cr]: Minor detail: "should" is just clutter

```suggestion
          it "returns the resulting array" do
```

File: src/compiler/crystal/macros/interpreter.cr [hunk 2]
```diff
@@ -76,13 +76,13 @@ module Crystal
...
                    @scope : Type, @path_lookup : Type, @location : Location?,
-                   @vars = {} of String => ASTNode, @block : Block? = nil, @def : Def? = nil,
+                   @vars = Hash(String | Int32, ASTNode).new, @block : Block? = nil, @def : Def? = nil,
                    @in_macro = false)
       @str = IO::Memory.new(512) # Can't be String::Builder because of `{{debug}}`
       @last = Nop.new
     end
 
-    def define_var(name, value)
+    def define_var(name : String | Int32, value : ASTNode) : Nil
```
Comment (straight-shoota) [inline src/compiler/crystal/macros/interpreter.cr]: Where does that `Int32` come from? You're adding calls to this method with `*_arg.name` which should surely be `String`.

File: src/compiler/crystal/macros/methods.cr [hunk 3]
```diff
@@ -2207,6 +2207,19 @@ private def interpret_array_or_tuple_method(object, klass, method, args, block,
         interpreter.accept block.body
       end
     end
+  when "map_with_index"
+    object.interpret_argless_method(method, args) do
+      raise "map_with_index expects a block" unless block
+
+      block_arg = block.args.first?
```
Comment (straight-shoota) [inline src/compiler/crystal/macros/methods.cr]: For clarity and similarity, I'd use `[0]?` here.

File: src/compiler/crystal/syntax/ast.cr [hunk 4]
```diff
@@ -339,6 +339,10 @@ module Crystal
       new(values.map { |value| (yield value).as(ASTNode) }, of: of)
     end
 
+    def self.map_with_index(values, of = nil)
```
Comment (straight-shoota) [inline src/compiler/crystal/syntax/ast.cr]: The `of` argument is never used. Is it necessary?

Comment (Blacksmoke16) [reply src/compiler/crystal/syntax/ast.cr]: It's used on the next line `new(..., of: of)`

Comment (straight-shoota) [reply src/compiler/crystal/syntax/ast.cr]: I mean it's never used as an argument to this method. You could just call `new(..., of: nil)`.

Comment (Blacksmoke16) [reply src/compiler/crystal/syntax/ast.cr]: Ah, gotcha

File: src/compiler/crystal/macros.cr [hunk 5]
```diff
@@ -422,6 +422,10 @@ module Crystal::Macros
     def gsub(regex : RegexLiteral, replacement : StringLiteral) : StringLiteral
     end
 
+    # Returns a `StringLiteral` where all `:` are replaced with `_`.
+    def identify : StringLiteral
```
Comment (RX14) [inline src/compiler/crystal/macros.cr]: This should be another PR?

**Comment by RX14** [comment]
What is the `StringLiteral#indentify` macro method useful for?

**Comment by Blacksmoke16** [comment]
@RX14 No idea, it already exists but just wasn't documented.

**Comment by Blacksmoke16** [comment]
Bump, this should be good to go.

**Comment by Blacksmoke16** [comment]
@RX14 @bcardiff @straight-shoota Any chance of this making in the next release?

**Comment by Blacksmoke16** [comment]
Anything left to do here?

**Comment by straight-shoota** [comment]
This needs to be washed into to commits, one focused on adding `map_with_index` and the other containing the unrelated changes.

**Comment by Blacksmoke16** [comment]
@straight-shoota Done.

**Comment by straight-shoota** [comment]
Yes, please. It should be a separate commit, either in this PR or in another one.
